### DISCUSSION TITLE
Importing new dashboards

### DISCUSSION THREAD

**Discussion by luddite516** [NetApp/harvest#819] [Q&A] [answered]
Hi:

I need some help with the 'harvest grafana import' process.  
We have two harvest/grafana appliances, one production and one test/dev.
I am trying to to export new dashboards from test/dev into prod.
I was able to get the 'export' command to work, and I now have the dashboard JSON files stored on my prod appliance.

When I try to use 'import' to pull the new dashboard into Grafana, I get the following error:

using API token from config
connected to Grafana server (version: 8.3.3)
folder [Utility] exists in Grafana - OK
folder [Harvest 21.11 - 7-mode] exists in Grafana - OK
preparing to import dashboards...
error importing [db_newview.json] - server response (404 - 404 Not Found) map[message:Dashboard not found status:not-found]

Not understanding how import works.  This appears to be saying that I can only import a dashboard which already exists in Grafana?

**Comment by cgrinds** [comment]
hi @luddite516 we'll take a look at the error - in the interim, you should be able to import/export directly from the Grafana UI also. Can you try importing your exported Json file by clicking `Import`
![image](https://user-images.githubusercontent.com/242252/152205017-ec7dd037-c006-48bb-9019-bdd89bb6fd8b.png)

then

![image](https://user-images.githubusercontent.com/242252/152205087-7743c31b-8a7e-4ae0-a535-2e70903e9307.png)

and selecting the exported json file?

**Comment by cgrinds** [comment]
And can you paste the exact import and export commands you used so we can try to recreate?

**Comment by luddite516** [comment]
Here is the export command from test/dev:

bin/harvest grafana export --addr https://XXXXXXXXXX --https --token 'XXXXXXXXXXXXXXX' --folder 'Utility' --folder-7mode 'Utility'

I run the above command on the prod server to connect to test/dev and pull its dashboards to prod.  That command succeeds.

Then, I run this command to try and pull the dashboards into Grafana on the prod server:

bin/harvest grafana import --addr https://XXXXXXXXX --datasource "Prometheus via Thanos" --directory grafana/dashboards --https --token 'XXXXXXX' --folder 'Utility' --directory-cdot grafana/dashboards/Utility
using API token from config
connected to Grafana server (version: 8.3.3)
folder [Utility] exists in Grafana - OK
folder [Harvest 21.11 - 7-mode] exists in Grafana - OK
preparing to import dashboards...
error importing [db_cigna_home.json] - server response (404 - 404 Not Found) map[message:Dashboard not found status:not-found

**Comment by cgrinds** [comment, marked as answer]
Thanks for the details @luddite516, I understand the bug now.
When exporting, Grafana also exports each dashboards's `id` 

![image](https://user-images.githubusercontent.com/242252/152214052-3883d3d1-c244-4cf7-b616-621692c84aff.png)

Harvest `import` uses the Grafana create/update dashboard [api](https://grafana.com/docs/grafana/latest/http_api/dashboard/#create--update-dashboard) to import the dashboards. That API says 
> dashboard.id – id = null to create a new dashboard.

but since the dashboards that you're trying to import have an `id`, Grafana will try to update instead of create. The update is failing because the dashboard does not exist to update. Grafana returns the `(404 - 404 Not Found) map[message:Dashboard not found status:not-found`

# Possible workarounds
- The Grafana import UI I pasted above handles this case better and will "do the right thing" when the dashboard does not exist
- Depending on how comfortable you are with gron or jq, you could write a small script to remove the id. Using gron as an example, something like this should do the trick `gron tmp_dash/luddite516_-_cDOT/db_netapp_detail_aggregate.json | grep -v "json.id =" | gron -u`

We'll work on a fix for this, targeted for mid month
